internal/engine/docker_compose_log_manager.go

```diff
@@ -0,0 +1,148 @@
+	if !state.WatchMounts {
+		return nil, nil
+	}
+
+	for _, ms := range state.ManifestStates {
+		if !ms.Manifest.IsDockerCompose() {
+			continue
+		}
+
+		existing, isActive := m.watches[ms.Manifest.Name]
+		startWatchTime := time.Unix(0, 0)
+		if isActive {
+			if existing.ctx.Err() == nil {
+				// Watcher is still active, no action needed.
+				continue
+			}
+
+			// The active log watcher got cancelled somehow, so we need to create
+			// a new one that picks up where it left off.
+			startWatchTime = <-existing.terminationTime
+		}
+
+		ctx, cancel := context.WithCancel(ctx)
+		w := DockerComposeLogWatch{
+			ctx:             ctx,
+			cancel:          cancel,
+			name:            ms.Manifest.Name,
+			dcYAMLPath:      ms.Manifest.DcYAMLPath,
+			startWatchTime:  startWatchTime,
+			terminationTime: make(chan time.Time, 1),
+		}
+		m.watches[ms.Manifest.Name] = w
+		setup = append(setup, w)
+	}
+
+	for key, value := range m.watches {
+		_, inState := state.ManifestStates[key]
+		if !inState {
+			delete(m.watches, key)
+
+			teardown = append(teardown, value)
+		}
+	}
+
+	return setup, teardown
+}
+
+func (m *DockerComposeLogManager) OnChange(ctx context.Context, st store.RStore) {
+	setup, teardown := m.diff(ctx, st)
+	for _, watch := range teardown {
+		watch.cancel()
+	}
+
+	for _, watch := range setup {
+		go m.consumeLogs(watch, st)
+	}
+}
+
+func (m *DockerComposeLogManager) consumeLogs(watch DockerComposeLogWatch, st store.RStore) {
+	defer func() {
+		watch.terminationTime <- time.Now()
+		watch.cancel()
+	}()
+
+	name := watch.name
+	readCloser, err := dockercompose.LogReaderForService(watch.ctx, watch.name.String(), watch.dcYAMLPath)
+	if err != nil {
+		logger.Get(watch.ctx).Infof("Error streaming %s logs: %v", name, err)
+		return
+	}
+	defer func() {
+		_ = readCloser.Close()
+	}()
+
+	// TODO(maia): docker-compose already prefixes logs, but maybe we want to roll
+	// our own (as in PodWatchManager) cuz it's prettier?
+	logWriter := logger.Get(watch.ctx).Writer(logger.InfoLvl)
+	actionWriter := DockerComposeLogActionWriter{
+		store:        st,
+		manifestName: name,
+	}
+	multiWriter := io.MultiWriter(logWriter, actionWriter)
+
+	_, err = io.Copy(multiWriter, NewHardCancelReader(watch.ctx, readCloser))
+	if err != nil && watch.ctx.Err() == nil {
+		logger.Get(watch.ctx).Infof("Error streaming %s logs: %v", name, err)
+		return
+	}
+}
+
+type DockerComposeLogWatch struct {
+	ctx             context.Context
+	cancel          func()
+	name            model.ManifestName
+	dcYAMLPath      string
+	startWatchTime  time.Time
+	terminationTime chan time.Time
+
+	// TODO(maia): do we need to track these? (maybe if we implement with `docker logs <cID>`...)
+	// cID             container.ID
+	// cName           container.Name
+}
+
+type DockerComposeLogActionWriter struct {
+	store        store.RStore
+	manifestName model.ManifestName
+}
+
+func (w DockerComposeLogActionWriter) Write(p []byte) (n int, err error) {
```

Comment:
Funky, I love it! Is this used for anything though?

Reply:
it's part of the multiWriter at ln 113 -- so when we copy to the multiWriter at ln 115 we're writing to both the output logger (i.e. allll the tilt logs) and the `DockerComposeLogActionWriter` (which, when we call `Write()`, dispatches a `DockerComposeLogAction` event)